packages/ethereum-contracts/test/foundry/utils/TOGA.t.sol

```diff
@@ -0,0 +1,483 @@
+     * @param account The address of the account to check.
+     * @param expectedNetFlow The expected net flow.
+     */
+    function _assetNetflow(
+        ISuperToken token,
+        address account,
+        int96 expectedNetFlow
+    ) internal {
+        int96 flowRate = sf.cfa.getNetFlow(token, account);
+        assertTrue(flowRate == expectedNetFlow);
+    }
+
+    function _deleteFlow(
+        ISuperToken token,
+        address sender,
+        address receiver
+    ) internal {
+        vm.startPrank(sender);
+        token.deleteFlow(sender, receiver);
+        vm.stopPrank();
+    }
+
+    function _startStream(
+        address sender,
+        address receiver,
+        int96 flowRate
+    ) internal {
+        vm.startPrank(sender);
+        superToken.createFlow(receiver, flowRate);
+        vm.stopPrank();
+    }
+
+    /**
+     * @dev Sends a PIC bid.
+     * @param sender The address of the sender.
+     * @param token The Super Token representing the asset.
+     * @param bond The bond amount.
+     * @param exitRate The exit rate.
+     */
+    function _sendPICBid(
+        address sender,
+        ISuperToken token,
+        uint256 bond,
+        int96 exitRate
+    ) internal {
+        vm.startPrank(sender);
+        token.send(address(toga), bond, abi.encode(exitRate));
+        vm.stopPrank();
+    }
+
+    function _sendPICBid(
+        address sender,
+        ISuperToken token,
+        uint256 bond,
+        bytes memory exitRate
+    ) internal {
+        vm.startPrank(sender);
+        token.send(address(toga), bond, exitRate);
+        vm.stopPrank();
+    }
+
+    function _shouldDefaultExitRate(
+        uint256 bondAmount
+    ) internal pure returns (int96) {
+        return int96(int256(bondAmount / (MIN_BOND_DURATION * 4)));
+    }
+
+    function _shouldMaxExitRate(
+        uint256 bondAmount
+    ) internal pure returns (int96) {
+        return int96(int256(bondAmount / MIN_BOND_DURATION));
+    }
+
+    function _collectRewards(
+        ISuperToken token,
+        int96 flowRate,
+        uint256 time
+    ) internal {
+        _startStream(admin, address(toga), flowRate);
+        vm.warp(block.timestamp + time);
+        vm.startPrank(admin);
+        token.deleteFlow(admin, address(toga));
+        vm.stopPrank();
+    }
+
+    // test
+
+    /**
+     * @dev Tests the contract setup.
+     */
+    function testContractSetup() public {
+        assertTrue(
+            toga.minBondDuration() == MIN_BOND_DURATION,
+            "minBondDuration"
+        );
+    }
+
+    /**
+     * @dev Tests that Alice becomes the PIC.
+     */
+    function testAliceBecomesPIC() public {
+        assertTrue(address(0) == toga.getCurrentPIC(superToken));
+
+        vm.startPrank(alice);
+        superToken.send(
+            address(toga),
+            BOND_AMOUNT_1E18,
+            abi.encode(EXIT_RATE_1)
+        );
+
+        assertTrue(toga.getCurrentPIC(superToken) == alice);
```

Comment:
missing `vm.expectEmit` testing that's present in typescript?

Reply:
Annoyingly, would need to copy the event signature in this contract

Reply:
Looking into it

Reply:
Added `vm.expectEmit` to test all the events.